pkg/storage/unified/resource/kv.go

```diff
@@ -0,0 +1,277 @@
+	SortOrderDesc
+)
+
+type ListOptions struct {
+	Sort     SortOrder
+	StartKey string
+	EndKey   string
+	Limit    int64
+}
+
+type GetOptions struct{}
+
+type KVObject struct {
+	Key   string
+	Value []byte
+}
+
+type KV interface {
+	// Keys returns all the keys in the store
+	Keys(ctx context.Context, section string, opt ListOptions) iter.Seq2[string, error]
+
+	// Get retrieves a key-value pair from the store
+	Get(ctx context.Context, section string, key string, opts ...GetOptions) (KVObject, error)
+
+	// List returns all the key-value pairs in the store
+	List(ctx context.Context, section string, opt ListOptions) iter.Seq2[KVObject, error]
+
+	// Save a new value
+	Save(ctx context.Context, section string, key string, value []byte) error
+
+	// Delete a value
+	Delete(ctx context.Context, section string, key string) error
+
+	// UnixTimestamp returns the current time in seconds since Epoch.
+	// This is used to ensure the server and client are not too far apart in time.
+	UnixTimestamp(ctx context.Context) (int64, error)
+}
+
+// Reference implementation of the KV interface using BadgerDB
+// This is only used for testing purposes, and will not work HA
+type badgerKV struct {
+	db *badger.DB
+}
+
+func NewBadgerKV(db *badger.DB) *badgerKV {
+	return &badgerKV{
+		db: db,
+	}
+}
+
+func (k *badgerKV) Get(ctx context.Context, section string, key string, opts ...GetOptions) (KVObject, error) {
+	txn := k.db.NewTransaction(false)
+	defer txn.Discard()
+
+	if section == "" {
+		return KVObject{}, fmt.Errorf("section is required")
+	}
+
+	key = section + "/" + key
+
+	item, err := txn.Get([]byte(key))
+	if err != nil {
+		if errors.Is(err, badger.ErrKeyNotFound) {
+			return KVObject{}, ErrNotFound
+		}
+		return KVObject{}, err
+	}
+	out := KVObject{
+		Key:   string(item.Key())[len(section)+1:],
+		Value: []byte{},
+	}
+	err = item.Value(func(val []byte) error {
+		out.Value = make([]byte, len(val))
+		copy(out.Value, val)
+		return nil
+	})
+	if err != nil {
+		return KVObject{}, err
+	}
+	return out, nil
+}
+
+func (k *badgerKV) Save(ctx context.Context, section string, key string, value []byte) error {
+	if section == "" {
+		return fmt.Errorf("section is required")
+	}
+
+	txn := k.db.NewTransaction(true)
+	defer txn.Discard()
+
+	key = section + "/" + key
+
+	err := txn.Set([]byte(key), value)
+	if err != nil {
+		return err
+	}
+	return txn.Commit()
+}
+
+func (k *badgerKV) Delete(ctx context.Context, section string, key string) error {
+	if section == "" {
+		return fmt.Errorf("section is required")
+	}
+
+	txn := k.db.NewTransaction(true)
+	defer txn.Discard()
+
+	key = section + "/" + key
+
+	err := txn.Delete([]byte(key))
+	if err != nil {
+		return err
+	}
+	return txn.Commit()
+}
+
+func (k *badgerKV) Keys(ctx context.Context, section string, opt ListOptions) iter.Seq2[string, error] {
+	if section == "" {
+		return func(yield func(string, error) bool) {
+			yield("", fmt.Errorf("section is required"))
+		}
+	}
```

Comment:
Suggestion: I think returning an explicit error here would be better than "hiding" it in the iterator. This fails the whole operation and not a distinct item in the iterator.